### PR TITLE
Fix get process architecture

### PR DESCRIPTION
**Pull request by nohwnd**
## Description
Fix code in GetProcessArchitecture that assumed that the process is the current process, and checked IntPtr to figure out the architecture quickly. Instead we rely on intPtr only when we know we are looking at the current process, and in other cases we use the native api. 

There is also more caching and fallback added when IsWow64Process2 fails because that api is not available on the system yet.

## Related issue
Related https://github.com/dotnet/sdk/issues/25421
It fixes part of the issue, but it looks like the DumpMinidump is recompiled somehow in dotnet sdk, and end up being unusable.

### CODE REVIEW

**Review by Evangelink — APPROVED** [review]
LGTM

**Review by MarcoRossignoli — APPROVED** [review]

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/ProcessHelper.cs [hunk 1]
```diff
@@ -64,9 +96,33 @@ private static bool IsArm64(int processId)
...
+                    // When we know this is not 64-bit operating system, then all processes are running as 32-bit, both
+                    // the current process and other processes.
+                    return PlatformArchitecture.X86;
+                }
+
+                var isWow64Process = NativeMethods.IsWow64Process(process.Handle, out var isWow64);
+                if (!isWow64Process)
+                {
+                    // Do nothing we cannot log errors here.
+                }
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/ProcessHelper.cs:113]: I'd prefer to keep it as is. I will for sure miss that `_` once we can log errors there.

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/NativeMethods.cs [hunk 2]
```diff
@@ -14,7 +14,14 @@ internal class NativeMethods
     public const ushort IMAGE_FILE_MACHINE_UNKNOWN = 0;
 
     [DllImport("kernel32.dll", SetLastError = true)]
-    public static extern bool IsWow64Process2(IntPtr process, out ushort processMachine, out ushort nativeMachine);
+    public static extern bool IsWow64Process2([In] IntPtr process, [Out] out ushort processMachine, [Out] out ushort nativeMachine);
+
+    // A pointer to a value that is set to TRUE if the process is running under WOW64.
+    // If the process is running under 32-bit Windows, the value is set to FALSE.
+    // If the process is a 64-bit application running under 64-bit Windows, the value is also set to FALSE.
+    [DllImport("kernel32.dll", SetLastError = true, CallingConvention = CallingConvention.Winapi)]
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply src/Microsoft.TestPlatform.PlatformAbstractions/common/NativeMethods.cs:22]: added.

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/ProcessHelper.cs [hunk 3]
```diff
@@ -23,37 +23,69 @@ public string GetCurrentProcessLocation()
...
-           ? PlatformArchitecture.ARM64
-           : PlatformArchitecture.X64
-       : PlatformArchitecture.X86;
-
-    private static bool IsArm64(int processId)
     {
+        if (_currentProcess.Id == processId)
+        {
+            // If we already cached the current process architecture, no need to figure it out again.
+            if (_currentProcessArchitecture != null)
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/ProcessHelper.cs:43]: added.

**Pull request merged**